Adi Peled 204818678 Noy Shabo 205983240

In [1]:
import pandas as pd
import numpy as np

### **TASK 1**

**1. Load the data into a Pandas Dataframe.**


In [5]:
df = pd.read_csv('mobile_price_1.csv')
print(df)

        id  battery_power  m_dep  mobile_wt  ...  memory  screen  wifi    price
0        0            842    0.6        188  ...    7568     LCD     g   751.64
1        1           1021    0.7        136  ...   53759   Touch  none   885.17
2        2            563    0.9        145  ...   41140   Touch  none  1067.78
3        3            615    0.8        131  ...   10558     LCD  none   819.22
4        4           1821    0.6        141  ...   44436   Touch  none  1023.13
...    ...            ...    ...        ...  ...     ...     ...   ...      ...
1995  1995            794    0.8        106  ...    2650   Touch  none   729.34
1996  1996           1965    0.2        187  ...   39573   Touch     n  1157.94
1997  1997           1911    0.7        108  ...   36706   Touch  none  1351.51
1998  1998           1512    0.1        145  ...   46218   Touch     g   846.52
1999  1999            510    0.9        168  ...   45731   Touch     a  1118.35

[2000 rows x 21 columns]


**2. Which of the categorical features are nominal and which are ordinal?**

answer:

nominal:
bluetooth,
screen


ordinal:
gen,
cores,
speed,
sim,
wifi

```

```



**3. Add a column that holds the total screen resolution for each device. Name it resolution.**

In [6]:
df['resolution']= df['px_height']*df['px_width']
print(df)

        id  battery_power  m_dep  mobile_wt  ...  screen  wifi    price  resolution
0        0            842    0.6        188  ...     LCD     g   751.64       15120
1        1           1021    0.7        136  ...   Touch  none   885.17     1799140
2        2            563    0.9        145  ...   Touch  none  1067.78     2167308
3        3            615    0.8        131  ...     LCD  none   819.22     2171776
4        4           1821    0.6        141  ...   Touch  none  1023.13     1464096
...    ...            ...    ...        ...  ...     ...   ...      ...         ...
1995  1995            794    0.8        106  ...   Touch  none   729.34     2309580
1996  1996           1965    0.2        187  ...   Touch     n  1157.94     1797975
1997  1997           1911    0.7        108  ...   Touch  none  1351.51     1416576
1998  1998           1512    0.1        145  ...   Touch     g   846.52      225120
1999  1999            510    0.9        168  ...   Touch     a  1118.35     

**4. Add a column that holds the DPI (dots per inch) of the screen width and name it DPI_w.**

In [7]:
df['DPI_w'] = np.where(df['sc_w'] == 0, np.nan, df['px_width']/(df['sc_w']*0.39370079))  # 1cm = 0.39370079 inches 
print(df)

        id  battery_power  m_dep  ...    price  resolution        DPI_w
0        0            842    0.6  ...   751.64       15120   274.319998
1        1           1021    0.7  ...   885.17     1799140  1683.173322
2        2            563    0.9  ...  1067.78     2167308  2179.319986
3        3            615    0.8  ...   819.22     2171776   567.054996
4        4           1821    0.6  ...  1023.13     1464096  1539.239990
...    ...            ...    ...  ...      ...         ...          ...
1995  1995            794    0.8  ...   729.34     2309580  1200.149992
1996  1996           1965    0.2  ...  1157.94     1797975   499.109997
1997  1997           1911    0.7  ...  1351.51     1416576  4145.279973
1998  1998           1512    0.1  ...   846.52      225120   170.179999
1999  1999            510    0.9  ...  1118.35      364182   478.789997

[2000 rows x 23 columns]


**5. Add a column that holds the ratio battery_power/talk_time and name it call_ratio.**

In [8]:
df['call_ratio']=np.where(df['talk_time'] == 0, np.nan,  df['battery_power']/df['talk_time'])
print(df)

        id  battery_power  m_dep  ...  resolution        DPI_w  call_ratio
0        0            842    0.6  ...       15120   274.319998         NaN
1        1           1021    0.7  ...     1799140  1683.173322  145.857143
2        2            563    0.9  ...     2167308  2179.319986   62.555556
3        3            615    0.8  ...     2171776   567.054996   55.909091
4        4           1821    0.6  ...     1464096  1539.239990  121.400000
...    ...            ...    ...  ...         ...          ...         ...
1995  1995            794    0.8  ...     2309580  1200.149992   41.789474
1996  1996           1965    0.2  ...     1797975   499.109997  122.812500
1997  1997           1911    0.7  ...     1416576  4145.279973  382.200000
1998  1998           1512    0.1  ...      225120   170.179999   79.578947
1999  1999            510    0.9  ...      364182   478.789997  255.000000

[2000 rows x 24 columns]


**6. Change the memory column to hold the memory in GB instead of MB.**

In [9]:
df['memory']= df['memory']/1024
print(df['memory'])

0        7.390625
1       52.499023
2       40.175781
3       10.310547
4       43.394531
          ...    
1995     2.587891
1996    38.645508
1997    35.845703
1998    45.134766
1999    44.659180
Name: memory, Length: 2000, dtype: float64


**7. Include the output of the `describe()` function of the dataframe.**

In [10]:
df.describe()

id  battery_power  ...        DPI_w   call_ratio
count  2000.000000    2000.000000  ...  1820.000000  1998.000000
mean    999.500000    1238.518500  ...   931.176477   166.621692
std     577.494589     439.418206  ...   992.678406   156.639198
min       0.000000     501.000000  ...    81.279999    25.100000
25%     499.750000     851.750000  ...   319.531998    75.750000
50%     999.500000    1226.000000  ...   536.857219   111.369318
75%    1499.250000    1615.250000  ...  1120.563326   186.068750
max    1999.000000    1998.000000  ...  5057.139967   989.500000

[8 rows x 18 columns]

**8. Convert the following features into categorical series in the Dataframe: speed,screen,cores**

In [11]:
# speed

sorted_speed = ['low', 'medium', 'high']
df['speed'] = pd.Categorical(df['speed'], ordered=True, categories=sorted_speed)


print(df['speed'])


0       medium
1          low
2          low
3         high
4       medium
         ...  
1995       low
1996      high
1997       low
1998       low
1999    medium
Name: speed, Length: 2000, dtype: category
Categories (3, object): ['low' < 'medium' < 'high']


In [12]:
# screen

types_of_screens = ['LCD', 'Touch']
df['screen'] = pd.Categorical(df['screen'], ordered=False, categories=types_of_screens)


print(df['screen'])


0         LCD
1       Touch
2       Touch
3         LCD
4       Touch
        ...  
1995    Touch
1996    Touch
1997    Touch
1998    Touch
1999    Touch
Name: screen, Length: 2000, dtype: category
Categories (2, object): ['LCD', 'Touch']


In [13]:
# cores

sorted_cores = ['Single', 'dual','triple','quad','penta','hexa','hepta','octa']
df['cores'] = pd.Categorical(df['cores'], ordered=True, categories=sorted_cores)


print(df['cores'])

0         dual
1       triple
2        penta
3         hexa
4         dual
         ...  
1995      hexa
1996      quad
1997      octa
1998     penta
1999      hexa
Name: cores, Length: 2000, dtype: category
Categories (8, object): ['Single' < 'dual' < 'triple' < 'quad' < 'penta' < 'hexa' < 'hepta' < 'octa']


### **TASK 2**

1. How many phones do not have a camera at all (front or back)?

In [21]:
df['not_camera_at_all']= df['f_camera'].isnull() & df['camera'].isnull() 
df['not_camera_at_all'].value_counts()


False    1899
True      101
Name: not_camera_at_all, dtype: int64

In [22]:
df=df.drop(['not_camera_at_all'],axis=1)


Answer=101 phones

**2. What is the average battery power for single-sim phones that have a camera or front camera with a higher resolution than 12 megapixels?**

In [23]:
filter= ((df['sim'] == 'Single') & ((df['camera'] >12) | (df['f_camera'] >12)  ))
only_true_rows=df[filter]
print(only_true_rows['battery_power'].mean())

1266.741935483871


Answer: mean= 1266.741935483871

**3. What is the ID and price of the most expensive phone that has no wifi, a touch screen and weighs more than 145 grams?**

In [24]:
filter2= (df['wifi']=='none') & (df['screen']=='Touch') & (df['mobile_wt']>145)
only_true_rows2=df[filter2]
max_price = only_true_rows2['price'].max()
filter_most_expensive = only_true_rows2['price'] == max_price
most_expensive =only_true_rows2[filter_most_expensive]
print(most_expensive['price'])

print('\n',most_expensive['id'] )


770    1533.85
Name: price, dtype: float64

 770    770
Name: id, dtype: int64


answer: id:770 , most_expensive price :  1533.85

**4. Create a pivot table that shows the percentage of phones with Bluetooth per generation, pivoted around the phone generation and split by “ram” quartiles. (i.e. the rows are the generation number and the columns are 4 quartiles of ram size).**

In [25]:
def changeSourceValueToDestValue(data,col,source_value1,dest_value1,source_value2, dest_value2):
    """get two source values and the requested convert values(dest_value) and col name and change the values from the sorce value to the dest value"""
    mapping={source_value1: dest_value1,source_value2:dest_value2}
    data[col]=data[col].map(mapping);
    return data;

df= changeSourceValueToDestValue(df, 'bluetooth', 'Yes',1, 'No',0)
ram_quartiles = pd.qcut(df['ram'],4)
print(df.pivot_table('bluetooth', ['gen'], [ram_quartiles]))

df= changeSourceValueToDestValue(df, 'bluetooth',1, 'Yes',0, 'No')

ram  (255.999, 1207.5]  (1207.5, 2146.5]  (2146.5, 3064.5]  (3064.5, 3998.0]
gen                                                                         
2             0.496063          0.529915          0.512605          0.552632
3             0.469565          0.500000          0.409091          0.443478
4             0.476744          0.486792          0.502008          0.538745


**5. Create a new Dataframe based on the original that has the following features: [id, battery_power, ram, talk_time, Bluetooth, cores, sim, memory, price], and contains a random sampling of half of the medium speed phones.**

In [26]:
filter_meduim_speed= df['speed']=='medium'
only_medium_rows=df[filter_meduim_speed]
half=only_medium_rows.sample(frac=0.5)

new_df = half.filter(['id', 'battery_power','ram','talk_time','bluetooth','cores','sim','memory','price'],axis=1)
print(new_df)

        id  battery_power   ram  talk_time  ...  cores     sim     memory    price
1076  1076            516   819         10  ...  penta    Dual  38.535156   684.05
404    404           1552  2403          9  ...   hexa  Single  13.211914  1120.21
28      28           1453  2373         10  ...   dual    Dual  51.240234   926.26
225    225           1634  3031          6  ...   dual  Single  17.579102  1027.44
578    578           1195   980         17  ...   quad    Dual  22.782227   777.38
...    ...            ...   ...        ...  ...    ...     ...        ...      ...
608    608           1975  2998         13  ...   octa  Single  15.560547  1339.15
1155  1155           1414  2290         12  ...   quad    Dual  48.299805   713.33
928    928           1349  2462         17  ...  hepta    Dual  54.354492   631.05
300    300           1937  3321         14  ...    NaN  Single  57.363281  1501.74
313    313            645   714         13  ...   octa  Single  40.167969   661.58

[41

**6. Using this new dataset, what is the maximum total talk time you can achieve if you use 3 phones, and which 3 phones will you use?**

In [27]:
sorted_phones = new_df.sort_values('talk_time',ascending=False)
print(sorted_phones)
print(sorted_phones['id'][0:3])
print('sum: ',sorted_phones['talk_time'][0:3].sum())

        id  battery_power   ram  talk_time  ...   cores     sim     memory    price
1181  1181           1271   892         20  ...     NaN    Dual   2.174805   655.30
1307  1307            831  1704         20  ...   penta    Dual  25.683594   785.60
1743  1743           1673  3094         20  ...   penta  Single  41.819336   805.59
1164  1164            860  2832         20  ...  triple    Dual  39.541992   925.36
1527  1527           1837  2294         20  ...    octa    Dual  12.182617  1211.55
...    ...            ...   ...        ...  ...     ...     ...        ...      ...
1562  1562           1325  2328          2  ...     NaN  Single  49.268555   953.45
656    656           1583  1704          2  ...   hepta    Dual  13.988281   806.67
925    925           1048  2146          2  ...     NaN  Single  29.227539   865.13
1365  1365           1318  1780          2  ...    dual    Dual  11.231445   927.81
1117  1117           1970  2372          2  ...    octa    Dual  57.366211  

answer: the three phones are the phones with id: 1621, 518, 1511 .

total talk time is: 60
